staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go

```diff
@@ -500,6 +535,14 @@ func (e *Store) Update(ctx context.Context, name string, objInfo rest.UpdatedObj
 		doUnconditionalUpdate := newResourceVersion == 0 && e.UpdateStrategy.AllowUnconditionalUpdate()
 
 		if existingResourceVersion == 0 {
+			if e.BeginCreate != nil {
+				fn, err := e.BeginCreate(obj, newCreateOptionsFromUpdateOptions(options))
+				if err != nil {
+					return nil, nil, err
+				}
+				cleanup = fn
```

Comment:
Overwriting `cleanup` here and calling it outside GuaranteedUpdate means that if GuaranteedUpdate is called multiple times, only the last invocation of BeginCreate will get its returned function invoked.

If we're allocating inside BeforeCreate, doesn't that mean:
* allocations of user-specified ports will fail on retries because they conflict with previously allocated things from previous invocations?
* auto-allocated ports from previous invocations will leak?

Reply:
this could be mitigated by calling `cleanup` on retry before invoking it again and overwriting `cleanup`, e.g.:

```go
if e.BeginCreate != nil {
  if cleanup != nil {
    // we were invoked previously, indicate that attempt did not succeed
    cleanup(false)
    cleanup = nil
  }
  fn, err := e.BeginCreate(obj, newCreateOptionsFromUpdateOptions(options))
  ...
  cleanup = fn
}
```

Reply:
basically, I would expect the invariant that every invocation of BeforeCreate/BeforeUpdate for a given request has its returned function called exactly once, and that call occurs prior to any reinvocations of BeforeCreate/BeforeUpdate for that request

Reply:
good point.  I did call it internally at first but then the confusion around AfterUpdate and Decorator hit me.  Before I make this more complex, let's sort that...

Reply:
Tests are green on additional commits to change other hooks to not return error.

Reply:
Now added a test for the inner-retry loop.